pages/building-packages-for-runtimes.md

```diff
@@ -0,0 +1,239 @@
+
+The build pipelines became so complex, it’s a proper combination of `bundlers`, `transpilers`, `node` and multiple module `formats`. Loading a `cjs` module from `npm` into project that is configuring to build as `esm` and vice-versa became so complex. At some point left most of the systems out of sync.
+
+`JSPM` builds all packages from `npm` ahead of time to spec complaint esm modules. And serves them using a distributed global CDN, regardless of the format in which the packages are published to npm. This makes it seamless for loading any module from `npm` and use it in any project at anytime.
+
+**Fields inside package.json**
+
+**browser**
+
+If a package is specifying [browser](https://docs.npmjs.com/cli/v9/configuring-npm/package-json#browser) field in the package.json. And when these packages are used in generating `import-maps` for browser as target. Then jspm uses this field instead of `main`, as the package authors explicitly stating the module that needs to be used when loading the same build inside a browser environments.
+
+**module**
+
+Module build detection in jspm. Let’s say, we are exposing `cjs` build from `main` and `esm` version in `module` field. And we don’t have `type` as `module` set on the package. Nor, the entry point is ending with `.mjs` extension. Something like below
+
+```json
+"main": "dist/index.js",
+"module": "dist/index.es.js"
+```
+
+This is one of the most used pattern in publishing packages. The builder uses `main` entry and builds the module into `esm` . This is because the `module` field is not in official specification but mostly used by `bundlers`. JSPM always follows the node specification.
+
+**exports**
+
+The package-builder from jspm parses the packages and creates an `export-map` for those packages which don’t expose them by itself. In the process, the builder is even intelligent enough to detect [subpaths](https://nodejs.org/api/packages.html#subpath-patterns) depending on the usage of the module imports internally. Which in return makes the package exports more efficient.
+
+> If the package is exposing an `exports` map, JSPM uses it instead of re-generating the whole export map.
+
+```json
+"exports": {
+    ".": {
+      "import": "./dist/default/lib.mjs",
+      "require": "./dist/default/lib.js",
+      "umd": "./dist/default/lib.umd.js"
+    }
+}
+```
+
+Let’s look it with an example with `react-router`. If we look into the `package.json` of the package
+
+[https://unpkg.com/browse/react-router@6.8.2/package.json](https://unpkg.com/browse/react-router@6.8.2/package.json)
+
+```json
+"main": "./dist/main.js",
+"unpkg": "./dist/umd/react-router.production.min.js",
+"module": "./dist/index.js",
+"types": "./dist/index.d.ts"
```

Comment:
Again, I would advise against top-level entries entirely for always using `"exports"`.

Reply:
Yup, totally get it. This is just straight away copy paste of their `package.json` from `unpkg`. Like to highlight how jspm processes and updates the `exports` field. I will highlight that it's a current state in `npm` something like that 👍